### ISSUE TITLE
How to create a "subscription" and only download new images?

### ISSUE THREAD

**Issue by TempMonkas** [Bionus/imgbrd-grabber#2246]
As title says. Currently I just have hundreds of entries in my downloads tab and run through them every so often, but it iters through every single file in each download - even ones I already have - and takes forever. Was wondering if there was a way to only download new stuff and skip the old ones. Currently it "skips" them but only after determining the file already exists.

If it's already possible, which I'm sure it is and I'm just an idiot who can't read, please link me the resource that explains how it works and/or ELI5 it for me, would be much appreciated!

**Comment by yami-no-tusbas** [comment]
First, there is the monitors to do this, but documentation is not up to date, you can still find information here #1620 
Second, for the "skip" part, the monitors does it automaticly, but you can always add the filter `"-grabber:downloaded"` as a tag to remove any downloaded file (the same as setting the save setting to : don't save) but the filter will only be applied to your monitor queries. (for exemple you can put a monitor on "pokemon -grabber:downloaded" and it'll should never download a file you already got, even if it was outside the monitors (like download job).

There no way to convert your download list to a monitors.json, but the simpliest way would be : 

1. to double clic the "tag" field in your download list
2. paste it on the search windows
3. put (or not) the "-grabber:downloaded" (could be usefull since the first time you add a monitors it has no "lasttime" to refer to, so it'll download a lot of already downloaded file, depending on your setting.)
4. then clic on "monitor" on the bottom right of the window.

A default monitor use default save path, etc, check new image every day, but everything is customizable by right click then select edit. There is currently a visual bug with editing that put the edited term at the bottom once you edited it so don't be surprised.

**Comment by TempMonkas** [comment]
> 
> 
> First, there is the monitors to do this, but documentation is not up to date, you can still find information here #1620
> Second, for the "skip" part, the monitors does it automaticly, but you can always add the filter `"-grabber:downloaded"` as a tag to remove any downloaded file (the same as setting the save setting to : don't save) but the filter will only be applied to your monitor queries. (for exemple you can put a monitor on "pokemon -grabber:downloaded" and it'll should never download a file you already got, even if it was outside the monitors (like download job).
> 
> There no way to convert your download list to a monitors.json, but the simpliest way would be :
> 
>     1. to double clic the "tag" field in your download list
> 
>     2. paste it on the search windows
> 
>     3. put (or not) the "-grabber:downloaded" (could be usefull since the first time you add a monitors it has no "lasttime" to refer to, so it'll download a lot of already downloaded file, depending on your setting.)
> 
>     4. then clic on "monitor" on the bottom right of the window.
> 
> 
> A default monitor use default save path, etc, check new image every day, but everything is customizable by right click then select edit. There is currently a visual bug with editing that put the edited term at the bottom once you edited it so don't be surprised.

Thanks, can you tell me where the download list file is stored? Curious to see if I can somehow automate it. I have nearly 1000 entries in it and really don't want to manually make monitors for them. Or perhaps this feature can be added in a future update?

**Comment by yami-no-tusbas** [comment]
Yes you can automate it i guess, it is a custom json file (with .igl extention), to access it you can simply use the menu File->Save downloads list. The monitor list is also json formatted. The monitors.json can be accessed in %AppData% -> Local -> Bionus -> Grabber

**Comment by Timurama** [comment]
cant you just use the tagging to filter out the results you don't need?

for example I use " id:>%id% sort:id:asc " on gelbooru to only show pictures that were posted after the file under the given id.

I'm sure there's a script like that on every booru if you look around in their respective tagging manuals.

**Comment by TempMonkas** [comment]
> 
> 
> cant you just use the tagging to filter out the results you don't need?
> 
> for example I use " id-:->%id% sort:-id-:asc " (with no hyphens) on gelbooru to only show pictures that were posted after the file under the given id.
> 
> I'm sure there's a script like that on every booru if you look around in their respective tagging manuals.

Well that would require me to do that manually for each tag. I want to simply have a master list that I can hit "Download" on which will only download new stuff without going through old stuff. Instead of taking hours it will take minutes.

**Comment by yami-no-tusbas** [comment]
Read the manual maybe, there is a "grabber:downloaded" filter, use "-" before it to reverse it (-grabber:downloaded), and you will not have what you already have downloaded, work even better if "keep md5s of deleted files" is checked.

**Comment by TempMonkas** [comment]
> 
> 
> Read the manual maybe, there is a "grabber:downloaded" filter, use "-" before it to reverse it (-grabber:downloaded), and you will not have what you already have downloaded, work even better if "keep md5s of deleted files" is checked.

This requires re-queuing the downloads, and my downloads are set up to go into folders based on the searched tag, so that will simply create a new folder. Can you maybe elaborate on what you mean? Maybe I'm missing something.

**Comment by yami-no-tusbas** [comment]
No need to re-queuing anything, assuming you have the download list, load it, then on the "post-filtering" column add for each : "-grabber:downloaded", and check in your settigns if grabber save md5, and keep md5 of deleted files, because if not grabber don't know it's downloaded. (Downloaded status is based on md5 AFAIK). You don't have to edit the search term, just add the post filter on each download. 

Or if you are confortable with it, use some notepadd++ or software alike (no office things...) and you can use the find & replace to change all the post-filter at once : (igl are json file after all, you can edit them)

Replace : 
```
"postFiltering": [
            ],
```
By :
```
"postFiltering": [
                "-grabber:downloaded"
            ],
```

After adding the filter, you can launch the download again, it should ignore the already existing files. The only problem is the maximum images and number of pages that can change over time, download list don't update theses at all.

For exemple, if an tag has 100 images on the first time you add it to the list, it will forever be view as 100 max images. So if on day one or two images are added, no problem, since they will be on the first page an push the other result further. But if for some reason, in one day you got more image than the "image limit" value, (unlikely I know) you may miss images because grabber will assume there is only a finite number of them.

Or if you need it converted send the download list to me by private message or something (I'm on Grabber discord), some regex may do the trick.

**Comment by TempMonkas** [comment]
> 
> 
> No need to re-queuing anything, assuming you have the download list, load it, then on the "post-filtering" column add for each : "-grabber:downloaded", and check in your settigns if grabber save md5, and keep md5 of deleted files, because if not grabber don't know it's downloaded. (Downloaded status is based on md5 AFAIK). You don't have to edit the search term, just add the post filter on each download.
> 
> Or if you are confortable with it, use some notepadd++ or software alike (no office things...) and you can use the find & replace to change all the post-filter at once : (igl are json file after all, you can edit them)
> 
> Replace :
> 
> ```
> "postFiltering": [
>             ],
> ```
> 
> By :
> 
> ```
> "postFiltering": [
>                 "-grabber:downloaded"
>             ],
> ```
> 
> After adding the filter, you can launch the download again, it should ignore the already existing files. The only problem is the maximum images and number of pages that can change over time, download list don't update theses at all.
> 
> For exemple, if an tag has 100 images on the first time you add it to the list, it will forever be view as 100 max images. So if on day one or two images are added, no problem, since they will be on the first page an push the other result further. But if for some reason, in one day you got more image than the "image limit" value, (unlikely I know) you may miss images because grabber will assume there is only a finite number of them.

Thanks so much for your informative reply! I've gotten the result I was looking for by applying the text replacement that you suggested in restore.igl. Although it takes longer for downloads to actually start since it's reading the MD5 list, it only attempts to download new images. 

Regarding what you said about the maximum images and number of pages issue, is there a way of fixing that? Perhaps of making those values not static in restore.igl? Or simply overstating the number to force it to check more pages that exist, so as to ensure it always sees the newest pages?

Thanks in advance!

**Comment by yami-no-tusbas** [comment]
Well a fix would be using the monitors instead of download queue. Since monitors don't care about how much image they have to download, but only the last date they checked for images.

Overstating won't work because at one time you may ask for a page who doesn't exist in the website, then returning error. Or you can make grabber go in a loop trying to get images that doesn't exist. Furthermore website may block you for spamming unseless request.

Plus it would be tedious to update thoses number since they can go up and down every time an image is added or removed, multiple times a minute for some popular tags. A "solution" could be to try to get a thousand image each time, putting a max 1000 so that you always check the last one thousand images. But still, since it'll have to check each md5's that not a good solution. Still the Monitors are better and since it is also json maybe you can create it with some kind of editor or a script that loop your "query" from the download list and construct a default monitor.json. I think I may be able to do it in php but it'll not be usefull for anyone without a local webserver.

**Comment by brazenvoid** [comment]
The download lists are best frankly as opposed to monitors in my experience. You have control over when you kick them off and can use any combination of score and other meta tags without fear. For me, I use the score meta tag to shear off the bulk of the results.

**Now to clear all your "worries":**

1. Grabber does not obey the max images, believe me I have many times discussed this with Jack and I am an old hoarder. That field is just for show. Meaning even if you modify it to 1000 on a 10k results tag, it will still go to the 10,000th image regardless. So that worry is inconsequential.
2. When I manage my download lists I have to add an age filter where possible to limit the downloads. Remember age filter does not work with Shimmie API (rule34.paheal.net etc).
3. As for images per page, this is dependent on source, Gelbooru API does not have a limit so you can safely set 1000 for every download regardless of the actual results.
4. -grabber:downloaded is not very performant as it checks the file on disk. If you do not change the names of your files then you can safely switch to -grabber:inMd5List which is much faster especially on hard disks. This also works to exclude files that you have once downloaded but now have deleted.

**Comment by TempMonkas** [comment]
> 
> 
> The download lists are best frankly as opposed to monitors in my experience. You have control over when you kick them off and can use any combination of score and other meta tags without fear. For me, I use the `score`, `height`, `width` and `size` meta tags to shear off the bulk of the results.
> 
> **Now to clear all your "worries":**
> 
>     1. Grabber does not obey the max images, believe me I have many times discussed this with Jack and I am an old hoarder. That field is just for show. Meaning even if you modify it to 1000 on a 10k results tag, it will still go to the 10,000th image regardless. So that worry is inconsequential.
> 
>     2. When I manage my download lists I have to add an age filter where possible to limit the downloads. Remember age filter does not work with Shimmie API (rule34.paheal.net etc). Usually I set it to `age:..1y` or `age:..3y`
> 
>     3. As for images per page, this is dependent on source, for Gelbooru API (gelbooru, danbooru, rule34.xxx etc) you can safely set 1000 for every download regardless of the actual results. Shimmie restricts to 70.
> 
>     4. `-grabber:downloaded` is not very performant as it checks the file on disk. You can safely switch to `-grabber:inMd5List` which is much faster especially on hard disks. This also works to exclude files that you have once downloaded but now have deleted outside grabber, otherwise those files will be downloaded yet again.

I have an issue with 4: both -grabber:downloaded and -grabber:inMd5List skip files that you already have, even if you only have them from another source. I already have Grabber set to hardlink already-existing images so that they can exist in multiple source folders without taking any additional space, so for me it seems these 2 arguments won't cut it.

Is there a way to still allow it to download images that exist in other source folders? eg. I sort my downloads by artist/source/image.jpg, so I'd want it to download an image to source1 even if it already exists in source2. I guess what I need is a check if it exists in the md5list AND the current directory?

**Comment by brazenvoid** [comment]
Essentially you want duplication. You can do this:

![Capture](https://user-images.githubusercontent.com/8722533/111752764-ebb78e00-88b7-11eb-8166-26fb38549cec.PNG)

And do not use the grabber post filters.

**Comment by brazenvoid** [comment]
You can also experiment changing your organization scheme by utilizing the new tags system available in nightly in concert with FileMeta. That way you will not need to organize these by search tags yourself. Usual windows search will get you to them.

**Comment by TempMonkas** [comment]
> 
> 
> Essentially you want duplication. You can do this:
> 
> ![Capture](https://user-images.githubusercontent.com/8722533/111752764-ebb78e00-88b7-11eb-8166-26fb38549cec.PNG)
> 
> And do not use the grabber post filters.

I thought I already mentioned that I already do this, except I hardlink the files so they don't take any additional space. Those two options are set to "Hard link". However, this makes downloading numerous jobs take forever since it needs to go through everything and determine whether or not I have it. And, if the name happened to have changed since the last time I downloaded it, it will download it in the same directory with the new name, thus creating a dupe in the same directory.

I don't want this. If the tags have changed since the last download, I don't want the file to be redownloaded if it already exists in the same directory. That's why I believe that a nice solution would be if I can have it check to see if it exists in the md5list AND if the md5 exists in the current directory.

> 
> 
> You can also experiment changing your organization scheme by utilizing the new tags system available in nightly in concert with FileMeta. That way you will not need to organize these by search tags yourself. Usual windows search with keywords will get you to them.

Thanks, but I'd really prefer keeping my current organization scheme.

**Comment by brazenvoid** [comment]
Well the program inherently relies on the md5 list which houses a hash with its location. This list has unique index for the hashes so essentially the program is not built to cater to your organization scheme. It is a design limitation so to speak. 

Through the settings I mentioned, it does allow dupes but in its database there is only one reference, the first one. The hard links are not registered in its database. So it will apply the same folder rule for the first one, all others will be ignored. 

Also similarly, there is no way around the dupe with different name, if you also want to allow dupes in other folders.

**Comment by brazenvoid** [comment]
You can use this post filter as already indicated, it will exclude the files on disk but will be slow for hundreds of files (in case of HDD): 
-grabber:downloaded 

These grabber* post filters were made to make browsing search results easier, not to optimize batch downloads.

**Comment by TempMonkas** [comment]
> 
> 
> You can use this post filter as already indicated, it will exclude the files on disk but will be slow for hundreds of files (in case of HDD):
> -grabber:downloaded
> 
> These grabber* post filters were made to make browsing search results easier, not to optimize batch downloads.

Ah, I see. That's unfortunate then, since the -grabber:downloaded post filter doesn't allow more than a single copy of the file, even if multiples should exist in other directories, defeating the purpose of the Hard Link setting I use. I added that command to all of my downloads not knowing this would be the case, so it looks like I'll have to now remove it.

I was really hoping the program would allow for AND/OR operators, so I could say, for example
"do not download if [is in md5 list] AND [md5 exists in current directory]"

You wouldn't know if that's possible, would you?

**Comment by brazenvoid** [comment]
As I said, the program only tracks the first reference, not the dupes. Your condition will be followed with my shown settings for the first file only.

You can exclude already existing images in the current destination folder with the post filter `grabber:alreadyExists`. This one does not cater the md5 list and matches the files manually. Perhaps it will be beneficial for you, if your internet connection is slow. Otherwise I am afraid, a fast connection with small files (200-300KB) will not put a dent on the overall job.

**Comment by brazenvoid** [comment]
I will also suggest making some sacrifices. Like:

- Adding some constraint on `minimum resolution`.
- Adding your most impactful blacklisted tags in your search query or post filter.
- Adding an `age` post filter (Only applies to Gelbooru API, Shimmie API errors with this).
- Adding a constraint on `score`.

I have been through all of the above, have reorganized many times already, downloading everything every time...

**Comment by brazenvoid** [comment]
Oh and the nightlies after v7.4.3, enforce the image limit defined in the batch downloads list. So you can use that too to limit your downloads. 

Though that will have a negative impact for images getting tagged later but perhaps you are not that much of a perfectionist hoarder :)

**Comment by TempMonkas** [comment]
> 
> 
> As I said, the program only tracks the first reference, not the dupes. Your condition will be followed with my shown settings for the first file only. The dupes will be allowed but will not be checked against for newer downloads in their folders.
> 
> You can exclude already existing images in the current destination folder with the post filter `grabber:alreadyExists`. This one does not cater the md5 list and matches the files manually. Perhaps it will be beneficial for you, if your internet connection is slow or metered. Otherwise I am afraid, a fast connection with small files (200-300KB) will not put a dent on the overall job.

Thank you very much for your assistance with this post and your 2 follow ups. I've done some experimenting and found that my preferred solution may be to simply use the date tag, such as "date:2021-01-01..", where in this case it only downloads anything posted after Jan 01, 2021.

This got me thinking, though: wouldn't it be great if the downloads page actually updated the date tag after each time you successfully download an entry? So it would update "date:2021-01-01.." to "date:2021-03-23.." if I were to use it right now. That way, dupes CAN still be downloaded, as desired for my use case, but it will not go through every old item that I've already downloaded, especially useful in helping prevent old images whose tags got updated to redownload.

I'll open a new issue and mention this there for coherence.

**Comment by brazenvoid** [comment]
Glad you figured out a solution for your organization scheme.

**Comment by TempMonkas** [comment]
> 
> 
> Glad you figured out a solution for your organization scheme.

Thanks, so am I. I hope my feature request thread I just posted is adequately clear. Do feel free to chime in on it.

**Comment by Bionus** [comment]
> I just come across this issue because I too would like not to redownload everything, simply because I've moved them elsewhere, and now Grabber thinks they needs redownloading becuase they no longer exists on disk.

Try the MD5 list fixer to have Grabber update its MD5 database for the new location, or enable "keep deleted files in MD5 list" to have it skip files from the MD5 list even if they've since been deleted/moved.

> Regardless, the filters age:..1y or age:..3y don't work. Always gives 0 results. Even literally on gelbooru.

AFAIK it only exists on Danbooru-based boards. Gelbooru doesn't support many search operators ([howto:cheatsheet](https://gelbooru.com/index.php?page=wiki&s=&s=view&id=26263)).

> grabber:downloaded always returns 0 results, just like grabber:inMd5List. So negating them will not have any effect. My MD5 list has tens of thousands of entries, and I know for a fact that from gelbooru I've download at least 1 image. Unless the MD5 list is somehow cycled.

Make sure you're using them as post-filters.
As for `grabber:alreadyExists` (so indirectly `grabber:downloaded`) there seem to be an issue in 7.5.1 that makes it not work properly.

> Something like date:2021-05-01.. does appear to filter something, but increasing the date to be very recent does not narrow the results in any way. I think what's happening is that it filters on images that happen to have any date at all, or something. So this filter is borderline useless.

If that's a post-filter, you won't see the impact until you actually reach the page in which the date changes for image results. Grabber can only post-filter results that are actually loaded. If there's 1K results, it can't know their dates otherwise.
